Hosting rework #3

HaoK opened the pull request:
Incorporated @lodejard feedback:

Introduce static WebApplication which is the static helpers used to create hosting engines and factories
- IHostingEngineFactory / IStartupLoader are the two hosting engine services which can be replaced
- IHostingEngine/HostingEngine are the actual workers which (is the merged result of Engine/Context)
- Nuked all of the wrapping service providers, made IHostingEnviroment get/set, and we now directly manipulate the instance.

cc @davidfowl @lodejard

davidfowl commented on src/Microsoft.AspNet.Hosting/HostingEngine.cs:
```diff
@@ -5,253 +5,340 @@
-            _applicationEnvironment = _fallbackServices.GetRequiredService<IApplicationEnvironment>();
-            _hostingEnvironment = new HostingEnvironment(_applicationEnvironment);
-            _fallbackServices = new WrappingServiceProvider(_fallbackServices, _hostingEnvironment, _appLifetime);
+            _config = config ?? new Configuration();
```
Dont allow nulls

davidfowl commented on src/Microsoft.AspNet.Hosting/IHostingEngine.cs:
```diff
@@ -0,0 +1,36 @@
+        // Accessing this will block Use methods
+        IServiceProvider ApplicationServices { get; }
+
+        // Use methods blow up after any of the above methods are called
```
Why don't we separate these?

HaoK replied:
We tried this, and decided against splitting builder/engine

davidfowl commented on src/Microsoft.AspNet.Hosting/IHostingEngine.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Microsoft.AspNet.Hosting
+{
+    public interface IHostingEngine
```
And why isn't this object IDisposable

davidfowl commented on src/Microsoft.AspNet.Hosting.Interfaces/IHostingEnvironment.cs:
```diff
@@ -7,10 +7,13 @@ namespace Microsoft.AspNet.Hosting
     public interface IHostingEnvironment
     {
+        // This must be settable!
         string EnvironmentName { get; set;  }
```
Can you change this in the application so that you start off as dev then end up as prod? Is that what we want?

HaoK commented:
Added support for `TestServer.Create<TStartup>` Will address PR feedback next

HaoK commented:
Added support for swapping out Appname and BasePath in TestServer

HaoK commented:
Added full permutations of overloads for WebApplication statics, cleans up music store and some test usage nicely

davidfowl commented on src/Microsoft.AspNet.Hosting/HostingEngine.cs:
```diff
@@ -5,251 +5,232 @@
-        private readonly IApplicationEnvironment _applicationEnvironment;
-        private readonly HostingEnvironment _hostingEnvironment;
+        // Start/ApplicationServices block use methods
+        private bool _useDisabled;
```
Do we even need this?

HaoK replied:
I can remove the guards, but then its up to the user to beware and know that calling Use after Start() or accessing ApplicationServices is ignored / no-op

HaoK commented:
Renamed WebApplication -> AspNetHosting and dropped Hosting from CreateFactory/Engine
Bury'd and uglified IHostingServicesBuilder to a super scary non interfaced service in Internal namespace: RootServiceCollectionInitalizer !!

davidfowl commented:
Omg, AspNetHosting? That's bizzare. Can we not put product names in the API please?

HaoK commented:
Hosting conflicts with the namespace, so we need to making it something like Hosting but slightly diff... Host?

davidfowl commented:
Microsoft.AspNet.Hosting.WebHost?

davidfowl commented:
WebHost.Create?

HaoK commented:
WebHost.CreateEngine/Factory  sounds good to me, lets go with that

davidfowl commented:
Also, the file is still called WebApplication.cs

davidfowl commented on src/Microsoft.AspNet.Hosting/Internal/AutoRequestServicesStartupFilter.cs:
```diff
@@ -14,7 +14,10 @@ public Action<IApplicationBuilder> Configure(IApplicationBuilder app, Action<IAp
             {
                 app.UseMiddleware<RequestServicesContainerMiddleware>();
-                next(builder);
+                if (next != null)
```
next can be null?

HaoK replied:
Probably was an edge case with start up filters/mangled app pipelines, but I hit this yes...

davidfowl replied:
It should never be null, this would be awful to code against.

HaoK replied:
Fair enuf, removed

HaoK commented:
Renamed WebHost

HaoK commented:
HostingEngine will now throw if Start() is called when a server wasn't configured via UseServer()

davidfowl commented on test/Microsoft.AspNet.Hosting.Tests/HostingEngineTests.cs:
```diff
@@ -21,55 +23,51 @@ public class HostingEngineTests : IServerFactory
-                ServerFactory = this,
-                ApplicationName = "Microsoft.AspNet.Hosting.Tests"
-            };
+            var engine = WebHost.CreateEngine()
```
this isn't the engine, call it server

HaoK commented:
Merged 77e2dc263f11655312d4c73bb8e22d7b6254d485
